runtime/eventlog.c

```diff
@@ -0,0 +1,401 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                 Stephen Dolan, University of Cambridge                 */
+/*                      Enguerrand Decorne, Tarides                       */
+/*                                                                        */
+/*   Copyright 2020 University of Cambridge                               */
+/*   Copyright 2020 Tarides                                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include "caml/alloc.h"
+#include "caml/eventlog.h"
+#include "caml/misc.h"
+#include "caml/memory.h"
+#include "caml/osdeps.h"
+
+
+#ifdef _WIN32
+#include <wtypes.h>
+#include <process.h>
+#elif defined(HAS_UNISTD)
+#include <unistd.h>
+#endif
+
+#ifdef HAS_MACH_ABSOLUTE_TIME
+#include <mach/mach_time.h>
+#elif HAS_POSIX_MONOTONIC_CLOCK
+#include <time.h>
+#endif
+
+#ifdef CAML_INSTR
+
+#define CTF_MAGIC 0xc1fc1fc1
+#define CAML_TRACE_VERSION 0x1
+
+struct ctf_stream_header {
+  uint32_t magic;
+  uint16_t caml_trace_version;
+  uint16_t stream_id;
+};
+
+static struct ctf_stream_header header = {
+  CTF_MAGIC,
+  CAML_TRACE_VERSION,
+  0
+};
+
+#pragma pack(1)
+struct ctf_event_header {
+  uint64_t timestamp;
+  uint32_t pid;
+  uint32_t id;
+};
+
+struct event {
+  struct ctf_event_header header;
+  uint16_t  phase; /* for GC events */
+  uint16_t  counter_kind; /* misc counter name */
+  uint8_t  alloc_bucket; /* for alloc counters */
+  uint64_t count; /* for misc counters */
+};
+
+#define EVENT_BUF_SIZE 4096
+struct event_buffer {
+  uintnat ev_generated;
+  struct event events[EVENT_BUF_SIZE];
+};
+
+static struct event_buffer* evbuf;
+
+int64_t time_counter(void)
```

Comment:
Technically, there is a way to handle clock frequency definition in the CTF metadata file:
https://diamon.org/ctf/
Part 8, Clocks, shows that one can embed frequency and offset information into a globally defined clock in the metadata file.
However this approach was discarded because at the time, it was meant to be merged in the main runtime, and I was concerned with growing every users runtime's size with more string data than required.
Indeed, to work, this would mean we need to generate a metadata file per tracing session, vs the current approach of one metadata file per runtime, meaning embedding more data in the runtime itself.
Now since the runtime is a separate one, it may be tolerable to embed such data, however still to the cost of a extra complication in initialization code.
I think the current approach is, while not optimal, is the simplest in design, and may be revisited at a later date once the needs and requirements of the instrumented runtime evolves. (be it for multicore, vanilla OCaml or any third party users.)
Also a big part the initialization code (for `mach_timebase_info` and `QueryPerformanceFrequency`) are static variables, so this may not be as big a problem as it seems at first glance.

Reply:
After asking the question above, I also noticed that frequency and offset are not enough for Linux, which needs a more complex calculation to get the time from `tv_sec` and `tv_nsec`.

I'm not too worried about the initialization code itself but more about the test that has to be done each time. That said, the current version is good enough for the time being.